feat(native): Insert into bucketed but unpartitioned Hive table

anandamideShakyan opened the pull request:
## Description
Addresses https://github.com/prestodb/presto/issues/25104 
Currently, Presto does not support INSERT INTO operations on bucketed but unpartitioned Hive tables. This limitation originates from a hard check in HiveWriterFactory:

https://github.com/prestodb/presto/blob/master/presto-hive/src/main/java/com/facebook/presto/hive/HiveWriterFactory.java#L480

## Motivation and Context

Supporting writes to bucketed unpartitioned Hive tables in Presto would improve compatibility and enhance Presto’s ability to handle modern Hive table layouts. It's a reasonable and useful feature for users who wish to leverage bucketing for performance optimizations even without partitioning.

## Impact

This change would align Presto’s behavior with the broader SQL-on-Hadoop ecosystem and remove an artificial limitation that may block valid use cases — particularly in data warehousing environments where bucketing is used independently of partitioning.

## Release Notes
```
== RELEASE NOTES ==

Hive Connector Changes

* Add support for INSERT into bucketed but unpartitioned Hive tables in Hive, including follow-up fixes for native validation and insert handling.

```

anandamideShakyan committed:
Insert into bucketed but unpartitioned Hive table

aditi-pandit commented:
@anandamideShakyan : Thanks for this PR. 

Have you tried this functionality with Prestissimo ? You might need https://github.com/facebookincubator/velox/pull/13283 as well for it.

anandamideShakyan commented:
@aditi-pandit Sure I will add the support in Prestissimo after https://github.com/facebookincubator/velox/pull/13283 is merged.

aditi-pandit commented:
@anandamideShakyan : Ther are failures in product tests. PTAL.

```
2025-05-18 19:49:10 INFO: [78 of 435] com.facebook.presto.tests.hive.TestHiveBucketedTables.testInsertIntoBucketedTables (Groups: )
2025-05-18 19:49:11 INFO: FAILURE     /    com.facebook.presto.tests.hive.TestHiveBucketedTables.testInsertIntoBucketedTables (Groups: ) took 1.1 seconds
2025-05-18 19:49:11 SEVERE: Failure cause:
java.lang.IllegalArgumentException: No mutable table instance found for name TableHandle{name=bucket_nation}
	at io.prestodb.tempto.fulfillment.table.TablesState.get(TablesState.java:64)
	at io.prestodb.tempto.fulfillment.table.TablesState.get(TablesState.java:48)
	at com.facebook.presto.tests.hive.TestHiveBucketedTables.testInsertIntoBucketedTables(TestHiveBucketedTables.java:173)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.testng.internal.invokers.MethodInvocationHelper.invokeMethod(MethodInvocationHelper.java:135)
	at org.testng.internal.invokers.TestInvoker.invokeMethod(TestInvoker.java:673)
	at org.testng.internal.invokers.TestInvoker.invokeTestMethod(TestInvoker.java:220)
	at org.testng.internal.invokers.MethodRunner.runInSequence(MethodRunner.java:50)
	at org.testng.internal.invokers.TestInvoker$MethodInvocationAgent.invoke(TestInvoker.java:945)
	at org.testng.internal.invokers.TestInvoker.invokeTestMethods(TestInvoker.java:193)
	at org.testng.internal.invokers.TestMethodWorker.invokeTestMethods(TestMethodWorker.java:146)
	at org.testng.internal.invokers.TestMethodWorker.run(TestMethodWorker.java:128)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
```

anandamideShakyan commented:
> @anandamideShakyan : Thanks for this PR.
> 
> Have you tried this functionality with Prestissimo ? You might need [facebookincubator/velox#13283](https://github.com/facebookincubator/velox/pull/13283) as well for it.

I have tried it on Prestissimo, with one coordinator and one worker. I created a table in hive schema and tpcds catalog using:
```
CREATE TABLE cars (
    id BIGINT,
    name VARCHAR,
    brand VARCHAR
)
WITH (
    format = 'PARQUET',
    bucketed_by = ARRAY['id'],
    bucket_count = 4
);
```
Inserted values:
```
INSERT INTO cars (id, name, brand) VALUES
  (1, 'Model S', 'Tesla'),
  (2, 'Civic', 'Honda'),
  (3, 'Mustang', 'Ford'),
  (4, 'A4', 'Audi');
```
Was able to see the entries on running the select query:

<img width="1668" alt="Screenshot 2025-06-18 at 1 07 59 PM" src="https://github.com/user-attachments/assets/e6e6d6b0-f7df-4370-a4ca-11d493a25dca" />

steveburnett commented:
Consider adding an example of how to use this new ability, or at least a mention that this is now possible for users to do and why it's useful (as you wrote in the Description), to the documentation.

anandamideShakyan commented:
Inserts into bucketed Hive tables using the C++ (Velox) worker were failing during finishInsert with:
<img width="1164" height="344" alt="Screenshot 2026-02-06 at 8 12 48 PM" src="https://github.com/user-attachments/assets/9536d11f-a0c8-431b-8384-d90b50a2ddaf" />

`VerifyException: computeFileNamesForMissingBuckets`


This happens because Presto’s Hive metadata layer assumes exactly one file per bucket per partition.
If any bucket does not produce a file, Presto attempts to synthesize “missing bucket” files during commit.

The Java worker never hits this path because it always creates one file per bucket, even when a bucket receives zero rows.

The Velox (C++) HiveDataSink, however, only created writers for buckets that actually received rows. When a bucket was empty, no writer → no file, causing Presto to think the bucket was missing and fail verification.

This is why inserts succeeded when data happened to hit all buckets, and failed otherwise.

**Fix**

The fix ensures that Velox creates one writer (and therefore one output file) per bucket, matching Java worker behavior and Presto’s expectations.

Specifically:

During HiveDataSink::splitInputRowsAndEnsureWriters(), we now pre-create writers for all buckets (for each partition, if partitioned).

This guarantees that every bucket produces exactly one file, even if it contains zero rows.

As a result, computeFileNamesForMissingBuckets() is never triggered and finishInsert succeeds.

**To Do**

- This is a Velox-side fix (C++ worker behavior).

- The original PR is in Presto, but the correct fix belongs in Velox, so a separate Velox PR is required. Will create velox PR soon.

- This change aligns C++ worker semantics with Java worker semantics and Hive’s bucketing contract.

With this fix locally, I am able to insert into bucketed hive tables with and without sidecar. I am now looking at resolving the unit test failure that came after these changes : https://github.com/prestodb/presto/pull/25115/

aditi-pandit commented:
@anandamideShakyan : Presto has a property hive.create-empty-bucket-files to control whether to create empty bucket files. Seems like this should always be false for native engine.

But in any case, doesn't Presto server create the missing buckets on the co-ordinator in the TableFinish logic and not in the worker ? I feel it should be on co-ordinator in TableFinish as its only after seeing all the worker files should we know which buckets are empty. The individual worker cannot make this decision. 

This error seems like a local problem between Hive and Presto on the co-ordinator. 

Please can you recheck if something else is missing.

anandamideShakyan commented on presto-native-sidecar-plugin/src/test/java/com/facebook/presto/sidecar/NativeSidecarPluginQueryRunner.java:
```diff
@@ -53,6 +54,7 @@ public static DistributedQueryRunner getQueryRunner()
         // Launch distributed runner.
         DistributedQueryRunner queryRunner = (DistributedQueryRunner) PrestoNativeQueryRunnerUtils.nativeHiveQueryRunnerBuilder()
                 .setCoordinatorSidecarEnabled(true)
+                .setHiveProperties(ImmutableMap.of("hive.create-empty-bucket-files", "false"))
```
Yes, I did run it. With hive.create-empty-bucket-files=true,  I get `internal error` on the UI. Attaching the log of the error:
```

2026-04-02T02:11:01.355-0600	ERROR	task-event-loop-6	com.facebook.presto.execution.StageExecutionStateMachine	Stage execution 20260402_081101_00003_tf4mg.0.0 failed
com.google.common.base.VerifyException
	at com.google.common.base.Verify.verify(Verify.java:102)
	at com.facebook.presto.hive.HiveMetadata.computeFileNamesForMissingBuckets(HiveMetadata.java:2075)
	at com.facebook.presto.hive.HiveMetadata.computePartitionUpdatesForMissingBuckets(HiveMetadata.java:2033)
	at com.facebook.presto.hive.HiveMetadata.finishInsertInternal(HiveMetadata.java:2222)
	at com.facebook.presto.hive.HiveMetadata.finishInsert(HiveMetadata.java:2201)
	at com.facebook.presto.spi.connector.classloader.ClassLoaderSafeConnectorMetadata.finishInsert(ClassLoaderSafeConnectorMetadata.java:497)
	at com.facebook.presto.metadata.MetadataManager.finishInsert(MetadataManager.java:1000)
	at com.facebook.presto.metadata.StatsRecordingMetadataManager.finishInsert(StatsRecordingMetadataManager.java:324)
	at com.facebook.presto.sql.planner.LocalExecutionPlanner.lambda$createTableFinisher$3(LocalExecutionPlanner.java:3664)
	at com.facebook.presto.operator.TableFinishOperator.getOutput(TableFinishOperator.java:290)
	at com.facebook.presto.operator.Driver.processInternal(Driver.java:448)
	at com.facebook.presto.operator.Driver.lambda$processFor$11(Driver.java:331)
	at com.facebook.presto.operator.Driver.tryWithLock(Driver.java:757)
	at com.facebook.presto.operator.Driver.processFor(Driver.java:324)
	at com.facebook.presto.execution.SqlTaskExecution$DriverSplitRunner.processFor(SqlTaskExecution.java:1078)
	at com.facebook.presto.execution.executor.PrioritizedSplitRunner.process(PrioritizedSplitRunner.java:165)
	at com.facebook.presto.execution.executor.TaskExecutor$TaskRunner.run(TaskExecutor.java:619)
	at com.facebook.presto.$gen.Presto_null__testversion____20260402_080859_460.run(Unknown Source)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
	at java.base/java.lang.Thread.run(Thread.java:840)

```

anandamideShakyan commented on presto-hive/src/test/java/com/facebook/presto/hive/TestHiveIntegrationSmokeTest.java:
```diff
@@ -908,6 +908,117 @@ public void testCreateTableNonSupportedVarcharColumn()
+                "bucket_count = 11 " +
+                ") ";
+
+        assertUpdate(createTable);
```
Done.

anandamideShakyan commented on presto-hive/src/test/java/com/facebook/presto/hive/TestHiveIntegrationSmokeTest.java:
```diff
@@ -908,6 +908,117 @@ public void testCreateTableNonSupportedVarcharColumn()
+        assertQuery("SELECT * from " + tableName, "VALUES ('a0', 'b0', 'c0'), ('a1', 'b1', 'c1')");
+
+        assertUpdate(session, "DROP TABLE " + tableName);
+        assertFalse(getQueryRunner().tableExists(session, tableName));
```
Applied

aditi-pandit commented on presto-hive/src/test/java/com/facebook/presto/hive/TestHiveIntegrationSmokeTest.java:
```diff
@@ -908,6 +908,123 @@ public void testCreateTableNonSupportedVarcharColumn()
+
+            assertEquals(computeActual("SELECT * from " + tableName).getRowCount(), 0);
+
+            // make sure that we will get one file per bucket regardless of writer count configured
```
How are you validating there is only one file ? You could use a query with $path hidden column for it. https://prestodb.io/docs/0.272/connector/hive.html#extra-hidden-columns

anandamideShakyan replied:
$path only exposes files that contain rows. Since this test inserts only two records, multiple records could hash to the same bucket and empty bucket files would not be visible through $path. To verify that all 11 bucket files were created, we'd need to inspect the table location directly (or insert data guaranteed to populate every bucket).

aditi-pandit commented:
Velox based PR review.

@anandamideShakyan : Please fix the test issue found.

**Summary**
The PR moves core behavior in the right direction (enabling inserts into bucketed, unpartitioned Hive tables) and adds broad test coverage, but I found one new reliability issue in the new smoke tests.

**Issues Found**
🟡 Suggestion: presto-hive/src/test/java/com/facebook/presto/hive/TestHiveIntegrationSmokeTest.java (around testBucketedTable(...), added block near lines ~1000–1040 in this diff)
testBucketedTable uses a fixed table name ("test_bucketed_table") and does not use try/finally cleanup. Since this helper runs repeatedly across formats/settings, a mid-test failure can leak the table and cascade failures into subsequent invocations.
Suggested fix: use a unique table name per invocation (as done in testEmptyBucketedTable) and wrap the method body in try/finally with DROP TABLE IF EXISTS.

**Positive Observations**
Good functional change in HiveWriterFactory removing the hard blocker for bucketed-unpartitioned inserts.
Nice expansion of coverage across hive integration, product tests, and native tests.
Good cleanup/robustness improvement already applied in testEmptyBucketedTable (unique name + finally cleanup).

anandamideShakyan commented:
@aditi-pandit I have fixed the above test issue. 

Another thing is that the insertion fails in native execution because C++ workers add .parquet extension to target file names (e.g., "000000_0_<'queryId'>.parquet") while Java's `getFileExtension()` returns empty string for PARQUET format, causing the coordinator to look for files without extensions (e.g., "000000_0_<'queryId'>"). The coordinator's string-based matching in HiveMetadata.computeFileNamesForMissingBuckets() fails when it receives file names with extensions from native workers, triggering a VerifyException. 
```
2026-05-18T15:47:46.461-0600	ERROR	SplitRunner-5-503	com.facebook.presto.execution.executor.TaskExecutor	Error processing Split 20260518_214746_00004_gfxuf.0.0.0.0-0  (start = 1.52748596380708E8, wall = 4 ms, cpu = 1 ms, wait = 0 ms, calls = 2)
com.google.common.base.VerifyException
	at com.google.common.base.Verify.verify(Verify.java:102)
	at com.facebook.presto.hive.HiveMetadata.computeFileNamesForMissingBuckets(HiveMetadata.java:2064)
	at com.facebook.presto.hive.HiveMetadata.computePartitionUpdatesForMissingBuckets(HiveMetadata.java:2022)
	at com.facebook.presto.hive.HiveMetadata.finishInsertInternal(HiveMetadata.java:2211)
	at com.facebook.presto.hive.HiveMetadata.finishInsert(HiveMetadata.java:2190)
	at com.facebook.presto.spi.connector.classloader.ClassLoaderSafeConnectorMetadata.finishInsert(ClassLoaderSafeConnectorMetadata.java:497)
	at com.facebook.presto.metadata.MetadataManager.finishInsert(MetadataManager.java:1008)
	at com.facebook.presto.metadata.StatsRecordingMetadataManager.finishInsert(StatsRecordingMetadataManager.java:324)
	at com.facebook.presto.sql.planner.LocalExecutionPlanner.lambda$createTableFinisher$3(LocalExecutionPlanner.java:3671)
	at com.facebook.presto.operator.TableFinishOperator.getOutput(TableFinishOperator.java:290)
	at com.facebook.presto.operator.Driver.processInternal(Driver.java:448)
	at com.facebook.presto.operator.Driver.lambda$processFor$11(Driver.java:331)
	at com.facebook.presto.operator.Driver.tryWithLock(Driver.java:757)
	at com.facebook.presto.operator.Driver.processFor(Driver.java:324)
	at com.facebook.presto.execution.SqlTaskExecution$DriverSplitRunner.processFor(SqlTaskExecution.java:1078)
	at com.facebook.presto.execution.executor.PrioritizedSplitRunner.process(PrioritizedSplitRunner.java:165)
	at com.facebook.presto.execution.executor.TaskExecutor$TaskRunner.run(TaskExecutor.java:619)
	at com.facebook.presto.$gen.Presto_null__testversion____20260518_214546_460.run(Unknown Source)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
	at java.base/java.lang.Thread.run(Thread.java:840)


2026-05-18T15:47:46.463-0600	ERROR	task-event-loop-6	com.facebook.presto.execution.StageExecutionStateMachine	Stage execution 20260518_214746_00004_gfxuf.0.0 failed
com.google.common.base.VerifyException
	at com.google.common.base.Verify.verify(Verify.java:102)
	at com.facebook.presto.hive.HiveMetadata.computeFileNamesForMissingBuckets(HiveMetadata.java:2064)
	at com.facebook.presto.hive.HiveMetadata.computePartitionUpdatesForMissingBuckets(HiveMetadata.java:2022)
	at com.facebook.presto.hive.HiveMetadata.finishInsertInternal(HiveMetadata.java:2211)
	at com.facebook.presto.hive.HiveMetadata.finishInsert(HiveMetadata.java:2190)
	at com.facebook.presto.spi.connector.classloader.ClassLoaderSafeConnectorMetadata.finishInsert(ClassLoaderSafeConnectorMetadata.java:497)
	at com.facebook.presto.metadata.MetadataManager.finishInsert(MetadataManager.java:1008)
	at com.facebook.presto.metadata.StatsRecordingMetadataManager.finishInsert(StatsRecordingMetadataManager.java:324)
	at com.facebook.presto.sql.planner.LocalExecutionPlanner.lambda$createTableFinisher$3(LocalExecutionPlanner.java:3671)
	at com.facebook.presto.operator.TableFinishOperator.getOutput(TableFinishOperator.java:290)
	at com.facebook.presto.operator.Driver.processInternal(Driver.java:448)
	at com.facebook.presto.operator.Driver.lambda$processFor$11(Driver.java:331)
	at com.facebook.presto.operator.Driver.tryWithLock(Driver.java:757)
	at com.facebook.presto.operator.Driver.processFor(Driver.java:324)
	at com.facebook.presto.execution.SqlTaskExecution$DriverSplitRunner.processFor(SqlTaskExecution.java:1078)
	at com.facebook.presto.execution.executor.PrioritizedSplitRunner.process(PrioritizedSplitRunner.java:165)
	at com.facebook.presto.execution.executor.TaskExecutor$TaskRunner.run(TaskExecutor.java:619)
	at com.facebook.presto.$gen.Presto_null__testversion____20260518_214546_460.run(Unknown Source)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
	at java.base/java.lang.Thread.run(Thread.java:840)

```

I've fixed this by changing the coordinator logic to extract and compare bucket numbers instead of exact file names, which works regardless of extension presence, but I'd like guidance on whether we should also align C++ to match Java's approach, or if the bucket-based matching solution is the preferred fix.

aditi-pandit commented:
@tdcmeehan : Please can you help review this PR.

anandamideShakyan commented:
I have added some checks to disable validation on connector-specific partitions which contains null buckettopartition value and fail validation. @tdcmeehan Please let me know if my changes are okay or if we can do it in a better way.

aditi-pandit commented:
@anandamideShakyan : Please can you add a release note and documentation for this issue.

@steveburnett

steveburnett reviewed (changes requested):
Please add documentation for this to the Presto documentation. Perhaps here:

https://github.com/prestodb/presto/blob/master/presto-docs/src/main/sphinx/connector/hive.rst

As I wrote in [this comment](https://github.com/prestodb/presto/pull/25139#issuecomment-3019265341): 

"_Consider adding an example of how to use this new ability, or at least a mention that this is now possible for users to do and why it's useful (as you wrote in the Description), to the documentation._"

steveburnett commented:
> @anandamideShakyan : Please can you add a release note and documentation for this issue.
> 
> @steveburnett

Thank you for adding the release note @anandamideShakyan!

anandamideShakyan commented:
@steveburnett I have added the documentation. PTAL and let me know if anything else is needed. Thanks.

steveburnett reviewed (dismissed):
LGTM! (docs) 

Pull branch, local doc build, looks good. Thanks!

tdcmeehan commented on presto-docs/src/main/sphinx/connector/hive.rst:
```diff
@@ -245,6 +245,33 @@ Property Name                                            Description
 
 .. _constructor: https://github.com/apache/hadoop/blob/02a9190af5f8264e25966a80c8f9ea9bb6677899/hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/conf/Configuration.java#L844-L875
 
+Bucketed unpartitioned tables
```
This is removing a prior limitation.  To me, the surprising thing is that Presto didn't support inserts into bucketed but not partitioned tables originally.  I don't think we need to call this out in the form of a doc--if there wasn't a limitations section before that clearly described that we didn't support this, then I don't think we should add now a specific call out to inserts into bucketed but not partitioned tables.  It should just work like a regular insert.

@steveburnett what do you think?  It reads slightly strange to call this out here.  It should just work, same as inserts into partitioned but not bucketed tables.

steveburnett replied:
> This is removing a prior limitation. To me, the surprising thing is that Presto didn't support inserts into bucketed but not partitioned tables originally. I don't think we need to call this out in the form of a doc--if there wasn't a limitations section before that clearly described that we didn't support this, then I don't think we should add now a specific call out to inserts into bucketed but not partitioned tables. It should just work like a regular insert.
> 
> @steveburnett what do you think? It reads slightly strange to call this out here. It should just work, same as inserts into partitioned but not bucketed tables.

@tdcmeehan, good point! I requested documentation for this feature in this PR in an earlier comment because I read the PR is something that a Presto user could not do before that they can do now, and the explanation in the PR (and the explanation in the added documentation) justified the benefit to me of adding an example to the doc. 

I do get your point about it feeling slightly strange. Is this former now-absent limitation a Hive-only limitation? Would you consider the code example better placed if this was removed from the Hive Connector doc page, and moved to https://prestodb.io/docs/current/sql/insert.html ?

tdcmeehan replied:
@steveburnett it is connector specific, so if it were to go anywhere, it would go into the Hive doc.  However, once again, because this is a _lifted limitation_, not a specific feature of inserts, I just wonder why bother explicitly calling it out?  It seems like the documentation should have been added on the limitation, not on our removal of that limitation.

steveburnett replied:
Okay. @anandamideShakyan, please delete the documentation.

anandamideShakyan replied:
Sure. Done.

steveburnett reviewed (dismissed):
LGTM! (docs)

Documentation removed per discussion.
